drivers/i3c/i3c_dw.c

```diff
@@ -0,0 +1,2335 @@
+	uint32_t thld_ctrl;
+
+	config->irq_config_func();
+
+	thld_ctrl = sys_read32(config->regs + QUEUE_THLD_CTRL);
+	thld_ctrl &= (~QUEUE_THLD_CTRL_RESP_BUF_MASK & ~QUEUE_THLD_CTRL_IBI_STS_MASK);
+	sys_write32(thld_ctrl, config->regs + QUEUE_THLD_CTRL);
+
+	thld_ctrl = sys_read32(config->regs + DATA_BUFFER_THLD_CTRL);
+	thld_ctrl &= ~DATA_BUFFER_THLD_CTRL_RX_BUF;
+	sys_write32(thld_ctrl, config->regs + DATA_BUFFER_THLD_CTRL);
+
+	sys_write32(INTR_ALL, config->regs + INTR_STATUS);
+
+	sys_write32(INTR_MASTER_MASK, config->regs + INTR_STATUS_EN);
+	sys_write32(INTR_MASTER_MASK, config->regs + INTR_SIGNAL_EN);
+}
+
+/**
+ * @brief Calculate the odd parity of a byte.
+ *
+ * This function calculates the odd parity of the input byte, returning 1 if the
+ * number of set bits is odd and 0 otherwise.
+ *
+ * @param p The byte for which odd parity is to be calculated.
+ *
+ * @return The odd parity result (1 if odd, 0 if even).
+ */
+static uint8_t odd_parity(uint8_t p)
+{
+	p ^= p >> 4;
+	p &= 0xf;
+	return (0x9669 >> p) & 1;
+}
+
+/**
+ * @brief Send Common Command Code (CCC).
+ *
+ * @see i3c_do_ccc
+ *
+ * @param dev Pointer to controller device driver instance.
+ * @param payload Pointer to CCC payload.
+ *
+ * @return @see i3c_do_ccc
+ */
+static int dw_i3c_do_ccc(const struct device *dev, struct i3c_ccc_payload *payload)
+{
+	const struct dw_i3c_config *config = dev->config;
+	struct dw_i3c_data *data = dev->data;
+	struct dw_i3c_xfer *xfer = &data->xfer;
+	struct dw_i3c_cmd *cmd;
+	int ret, i, pos;
+	uint32_t present_state;
+
+	present_state = sys_read32(config->regs + PRESENT_STATE);
+	if (!(present_state & PRESENT_STATE_CURRENT_MASTER)) {
+		return -EACCES;
+	}
+
+	ret = k_mutex_lock(&data->mt, K_MSEC(1000));
+	if (ret) {
+		LOG_DBG("%s: Mutex err (%d)", dev->name, ret);
+		return ret;
+	}
+
+	memset(xfer, 0, sizeof(struct dw_i3c_xfer));
+	xfer->ret = -1;
+
+	/* in the case of multiple targets in a CCC, each command queue must have the same CCC ID
+	 * loaded along with different dev index fields pointing to the targets
+	 */
+	if (i3c_ccc_is_payload_broadcast(payload)) {
+		xfer->ncmds = 1;
+		cmd = &xfer->cmds[0];
+		cmd->buf = payload->ccc.data;
+
+		cmd->cmd_hi = COMMAND_PORT_ARG_DATA_LEN(payload->ccc.data_len) |
+			      COMMAND_PORT_TRANSFER_ARG;
+		cmd->cmd_lo = COMMAND_PORT_CP | COMMAND_PORT_TOC | COMMAND_PORT_ROC |
+			      COMMAND_PORT_CMD(payload->ccc.id);
+
+		if ((payload->targets.payloads) && (payload->targets.payloads[0].rnw)) {
+			cmd->cmd_lo |= COMMAND_PORT_READ_TRANSFER;
+			cmd->rx_len = payload->ccc.data_len;
+		} else {
+			cmd->tx_len = payload->ccc.data_len;
+		}
+	} else {
+		if (!(payload->targets.payloads)) {
+			LOG_ERR("%s: Direct CCC Payload structure Empty", dev->name);
+			ret = -EINVAL;
+			goto error;
+		}
+		xfer->ncmds = payload->targets.num_targets;
+		for (i = 0; i < payload->targets.num_targets; i++) {
+			cmd = &xfer->cmds[i];
+			pos = get_i3c_addr_pos(dev, payload->targets.payloads[i].addr);
```

Comment:
There was an issue found here where if the `addr` is a static addr in the case of SETDASA, then it won't return the correct `pos` as it does the lookup based on the dynamic addr. This will need to be fixed when I can get some time

Reply:
Spotted this while testing the driver on our side (currently on hold due to some issues we're investigating)

Reply:
> Spotted this while testing the driver on our side (currently on hold due to some issues we're investigating)

Ah, yeah :upside_down_face: . Was there anything else that you spotted?

Reply:
Not really, but due to some other HW issues, I haven't been able to test other advanced features.

Reply:
ah... we I pushed up a fix for this since then